File: frontend/src/services/api/import-export.ts
/**
 * Import/Export API Service
 *
 * Handles import and export of Heimdall system data:
 * - Export data to .heimdall JSON files
 * - Import data from .heimdall JSON files
 * - Get metadata about exportable data
 */

import api from '@/lib/api';

// Type definitions matching backend models

export interface CreatorInfo {
  username: string;
  name?: string;
}

export interface SectionSizes {
  settings: number;
  sources: number;
  websdrs: number;
  sessions: number;
  sample_sets: number;
  models: number;
  audio_library: number;
}

export interface ExportMetadata {
  version: string;
  created_at: string;
  creator: CreatorInfo;
  section_sizes: SectionSizes;
  description?: string;
}

export interface ExportedSource {
  id: string;
  name: string;
  description?: string;
  frequency_hz: number;
  latitude: number;
  longitude: number;
  power_dbm?: number;
  source_type?: string;
  is_validated: boolean;
  error_margin_meters?: number;
  created_at: string;
  updated_at: string;
}

export interface ExportedWebSDR {
  id: string;
  name: string;
  url: string;
  location_description?: string;
  latitude: number;
  longitude: number;
  altitude_meters?: number;
  country?: string;
  operator?: string;
  is_active: boolean;
  timeout_seconds: number;
  retry_count: number;
  created_at: string;
  updated_at: string;
}

export interface ExportedSession {
  id: string;
  known_source_id: string;
  session_name: string;
  session_start: string;
  session_end?: string;
  duration_seconds?: number;
  celery_task_id?: string;
  status: string;
  approval_status: string;
  notes?: string;
  created_at: string;
  updated_at: string;
  measurements_count: number;
}

export interface ExportedSampleSet {
  id: string;
  name: string;
  description?: string;
  num_samples: number;
  config?: Record<string, unknown>;
  quality_metrics?: Record<string, unknown>;
  created_at: string;
  samples?: Record<string, unknown>[];
}

export interface ExportedModel {
  id: string;
  model_name: string;
  version: number;
  model_type: string;
  created_at: string;
  onnx_model_base64?: string;
  accuracy_meters?: number;
  hyperparameters?: Record<string, unknown>;
  training_metrics?: Record<string, unknown>;
}

export interface UserSettings {
  theme: string;
  default_frequency_mhz: number;
  default_duration_seconds: number;
  auto_approve_sessions: boolean;
}

export interface ExportSections {
  settings?: UserSettings;
  sources?: ExportedSource[];
  websdrs?: ExportedWebSDR[];
  sessions?: ExportedSession[];
  sample_sets?: ExportedSampleSet[];
  models?: ExportedModel[];
  audio_library?: ExportedAudioLibrary[];
}

export interface HeimdallFile {
  metadata: ExportMetadata;
  sections: ExportSections;
}

export interface ExportRequest {
  creator: CreatorInfo;
  description?: string;
  include_settings?: boolean;
  include_sources?: boolean;
  include_websdrs?: boolean;
  include_sessions?: boolean;
  sample_set_configs?: SampleSetExportConfig[] | null;
  model_ids?: string[] | null;
  audio_library_ids?: string[] | null;
}

export interface ExportResponse {
  file: HeimdallFile;
  size_bytes: number;
}

export interface ImportRequest {
  heimdall_file: HeimdallFile;
  import_settings?: boolean;
  import_sources?: boolean;
  import_websdrs?: boolean;
  import_sessions?: boolean;
  import_sample_sets?: boolean;
  import_models?: boolean;
  import_audio_library?: boolean;
  overwrite_existing?: boolean;
}

export interface ImportResponse {
  success: boolean;
  message: string;
  imported_counts: Record<string, number>;
  errors: string[];
}

export interface AvailableSampleSet {
  id: string;
  name: string;
  num_samples: number;
  num_iq_samples: number;
  created_at: string;
  estimated_size_bytes: number;
  estimated_size_per_feature: number;
  estimated_size_per_iq: number;
}

export interface SampleSetExportConfig {
  dataset_id: string;
  sample_offset: number;
  sample_limit: number | null;
}

export interface AvailableModel {
  id: string;
  model_name: string;
  version: number;
  created_at: string;
  has_onnx: boolean;
}

export interface ExportedAudioChunk {
  id: string;
  chunk_index: number;
  duration_seconds: number;
  sample_rate: number;
  num_samples: number;
  file_size_bytes: number;
  original_offset_seconds: number;
  rms_amplitude?: number;
  created_at: string;
  audio_data_base64?: string;
}

export interface ExportedAudioLibrary {
  id: string;
  filename: string;
  category: string;
  tags?: string[];
  file_size_bytes: number;
  duration_seconds: number;
  sample_rate: number;
  channels: number;
  audio_format: string;
  processing_status: string;
  total_chunks: number;
  enabled: boolean;
  created_at: string;
  updated_at: string;
  chunks?: ExportedAudioChunk[];
}

export interface AvailableAudioLibrary {
  id: string;
  filename: string;
  category: string;
  duration_seconds: number;
  total_chunks: number;
  file_size_bytes: number;
  chunks_total_bytes: number;
  created_at: string;
}

export interface MetadataResponse {
  sources_count: number;
  websdrs_count: number;
  sessions_count: number;
  sample_sets: AvailableSampleSet[];
  models: AvailableModel[];
  audio_library: AvailableAudioLibrary[];
  estimated_sizes: SectionSizes;
}

/**
 * Get metadata about available data for export
 */
export async function getExportMetadata(): Promise<MetadataResponse> {
  const response = await api.get('/import-export/export/metadata');
  return response.data;
}

/**
 * Export selected data sections to .heimdall file format
 */
export async function exportData(request: ExportRequest): Promise<ExportResponse> {
  const response = await api.post('/import-export/export', request);
  return response.data;
}

/**
 * Import data from .heimdall file format
 */
export async function importData(request: ImportRequest): Promise<ImportResponse> {
  const response = await api.post('/import-export/import', request);
  return response.data;
}

/**
 * Download .heimdall file to user's computer (browser)
 */
export function downloadHeimdallFile(file: HeimdallFile, filename?: string): void {
  const json = JSON.stringify(file, null, 2);
  const blob = new Blob([json], { type: 'application/json' });
  const url = URL.createObjectURL(blob);

  const a = document.createElement('a');
  a.href = url;
  a.download = filename || `heimdall-export-${new Date().toISOString().split('T')[0]}.heimdall`;
  document.body.appendChild(a);
  a.click();
  
  // Cleanup after a short delay to ensure download starts
  setTimeout(() => {
    try {
      if (a.parentNode === document.body) {
        document.body.removeChild(a);
      }
    } catch (error) {
      // Silent fail - element already removed or not in body
      console.debug('Download link cleanup: already removed');
    }
    URL.revokeObjectURL(url);
  }, 100);
}

/**
 * Load .heimdall file from user's computer (browser)
 */
export function loadHeimdallFile(): Promise<HeimdallFile> {
  return new Promise((resolve, reject) => {
    const input = document.createElement('input');
    input.type = 'file';
    input.accept = '.heimdall,.json';

    let fileSelected = false;

    input.onchange = (e: Event) => {
      fileSelected = true;
      const target = e.target as HTMLInputElement;
      const file = target.files?.[0];

      if (!file) {
        reject(new Error('No file selected'));
        return;
      }

      const reader = new FileReader();

      reader.onload = (readerEvent: ProgressEvent<FileReader>) => {
        try {
          const content = readerEvent.target?.result as string;
          const heimdallFile = JSON.parse(content) as HeimdallFile;
          resolve(heimdallFile);
        } catch (_error) {
          reject(new Error('Invalid .heimdall file format'));
        }
      };

      reader.onerror = () => {
        reject(new Error('Error reading file'));
      };

      reader.readAsText(file);
    };

    // Handle cancel - user closed dialog without selecting a file
    input.oncancel = () => {
      reject(new Error('File selection cancelled'));
    };

    // Fallback for browsers that don't support oncancel
    setTimeout(() => {
      if (!fileSelected && !input.files?.length) {
        reject(new Error('File selection cancelled'));
      }
    }, 100);

    input.click();
  });
}

// ===== ASYNC EXPORT WITH STREAMING PROGRESS =====

/**
 * WebSocket event types for export progress
 */
export interface ExportProgressEvent {
  event: 'export:progress';
  task_id: string;
  stage: 'settings' | 'sources' | 'websdrs' | 'sample_sets' | 'finalizing';
  current: number;
  total: number;
  message: string;
  percentage?: number;
}

export interface ExportCompletedEvent {
  event: 'export:completed';
  task_id: string;
  status: 'completed' | 'failed';
  download_url?: string;
  file_size_bytes?: number;
  error_message?: string;
}

export type ExportWebSocketEvent = ExportProgressEvent | ExportCompletedEvent;

/**
 * Response from starting an async export
 */
export interface AsyncExportResponse {
  task_id: string;
  message: string;
}

/**
 * Start an async export job (returns immediately with task_id)
 * Progress is tracked via WebSocket events
 */
export async function exportDataAsync(request: ExportRequest): Promise<AsyncExportResponse> {
  const response = await api.post('/import-export/export/async', request);
  return response.data;
}

/**
 * Download exported file from MinIO via task_id
 * This should be called after receiving export:completed event
 */
export function downloadExportedFile(taskId: string): void {
  const downloadUrl = `/api/import-export/download/${taskId}`;
  
  // Use window.location for automatic download
  const a = document.createElement('a');
  a.href = downloadUrl;
  a.download = `heimdall-export-${taskId.slice(0, 8)}.heimdall`;
  document.body.appendChild(a);
  a.click();
  
  // Cleanup
  setTimeout(() => {
    if (a.parentNode === document.body) {
      document.body.removeChild(a);
    }
  }, 100);
}

File: services/backend/src/models/import_export.py
"""Import/Export data models for Heimdall SDR.

Defines the .heimdall file format for saving and restoring system state.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class CreatorInfo(BaseModel):
    """Information about who created the export."""

    username: str = Field(..., description="Username of the creator")
    name: str | None = Field(None, description="Full name of the creator")


class SectionSizes(BaseModel):
    """Byte sizes of each section in the export."""

    settings: int = 0
    sources: int = 0
    websdrs: int = 0
    sessions: int = 0
    sample_sets: int = 0
    models: int = 0
    audio_library: int = 0


class ExportMetadata(BaseModel):
    """Metadata about the exported file."""

    version: str = Field(default="1.0", description="File format version")
    created_at: datetime = Field(default_factory=datetime.utcnow, description="Export timestamp")
    creator: CreatorInfo
    section_sizes: SectionSizes = Field(default_factory=SectionSizes)
    description: str | None = Field(None, description="Optional description of export")


class ExportedSource(BaseModel):
    """Known RF source for export."""

    id: str
    name: str
    description: str | None = None
    frequency_hz: int
    latitude: float
    longitude: float
    power_dbm: float | None = None
    source_type: str | None = None
    is_validated: bool = False
    error_margin_meters: float | None = None
    created_at: str
    updated_at: str


class ExportedWebSDR(BaseModel):
    """WebSDR configuration for export."""

    id: str
    name: str
    url: str
    location_description: str | None = None
    latitude: float
    longitude: float
    altitude_meters: float | None = None
    country: str | None = None
    operator: str | None = None
    is_active: bool = True
    timeout_seconds: int = 30
    retry_count: int = 3
    created_at: str
    updated_at: str


class ExportedSession(BaseModel):
    """Recording session for export."""

    id: str
    known_source_id: str
    session_name: str
    session_start: str
    session_end: str | None = None
    duration_seconds: float | None = None
    celery_task_id: str | None = None
    status: str
    approval_status: str = "pending"
    notes: str | None = None
    created_at: str
    updated_at: str
    measurements_count: int = 0


class ExportedModel(BaseModel):
    """ML model metadata for export."""

    id: str
    model_name: str
    version: int
    model_type: str
    created_at: str
    # ONNX model binary encoded as base64
    onnx_model_base64: str | None = None
    # Model metrics and configuration
    accuracy_meters: float | None = None
    hyperparameters: dict | None = None
    training_metrics: dict | None = None


class UserSettings(BaseModel):
    """User settings for export."""

    theme: str = "light"
    default_frequency_mhz: float = 145.5
    default_duration_seconds: float = 10.0
    auto_approve_sessions: bool = False
    # Add more settings as needed


class IQData(BaseModel):
    """IQ data for a single receiver."""
    
    receiver_id: str  # e.g., "RX_000"
    iq_data_base64: str  # numpy array encoded as base64
    

class ExportedIQSample(BaseModel):
    """Complete IQ sample data from synthetic_iq_samples table."""
    
    id: str
    sample_idx: int
    timestamp: str
    tx_lat: float
    tx_lon: float
    tx_alt: float
    tx_power_dbm: float
    frequency_hz: int
    num_receivers: int
    gdop: float | None = None
    mean_snr_db: float | None = None
    overall_confidence: float | None = None
    receivers_metadata: list[dict]  # JSONB array of receiver configs
    iq_metadata: dict  # JSONB (duration_ms, sample_rate_hz, center_frequency_hz)
    iq_storage_paths: dict  # JSONB (mapping receiver_id -> S3 path)
    iq_data: list[IQData]  # Actual IQ binary data in base64
    created_at: str


class ExportedFeature(BaseModel):
    """Complete feature data from measurement_features table."""
    
    recording_session_id: str
    timestamp: str
    receiver_features: list[dict]  # JSONB array
    tx_latitude: float | None = None
    tx_longitude: float | None = None
    tx_altitude_m: float | None = None
    tx_power_dbm: float | None = None
    tx_known: bool = False
    extraction_metadata: dict  # JSONB
    overall_confidence: float
    mean_snr_db: float | None = None
    num_receivers_detected: int | None = None
    gdop: float | None = None
    extraction_failed: bool = False
    error_message: str | None = None
    created_at: str
    num_receivers_in_sample: int | None = None


class ExportedSampleSet(BaseModel):
    """Synthetic dataset (sample set) for export with complete data."""

    id: str
    name: str
    description: str | None = None
    num_samples: int
    config: dict | None = None
    quality_metrics: dict | None = None
    created_at: str
    # Complete feature data from measurement_features
    features: list[ExportedFeature] | None = None
    # Complete IQ data from synthetic_iq_samples
    iq_samples: list[ExportedIQSample] | None = None
    # Range information for partial exports
    num_exported_features: int | None = None
    num_exported_iq_samples: int | None = None
    export_range: dict | None = None  # {"offset": int, "limit": int}


class ExportedAudioChunk(BaseModel):
    """Audio chunk for export (1-second preprocessed audio)."""

    id: str
    chunk_index: int
    duration_seconds: float
    sample_rate: int
    num_samples: int
    file_size_bytes: int
    original_offset_seconds: float
    rms_amplitude: float | None = None
    created_at: str
    # Audio data encoded as base64
    audio_data_base64: str | None = None


class ExportedAudioLibrary(BaseModel):
    """Audio library entry for export."""

    id: str
    filename: str
    category: str
    tags: list[str] | None = None
    file_size_bytes: int
    duration_seconds: float
    sample_rate: int
    channels: int
    audio_format: str
    processing_status: str
    total_chunks: int
    enabled: bool
    created_at: str
    updated_at: str
    # Associated audio chunks
    chunks: list[ExportedAudioChunk] | None = None


class ExportSections(BaseModel):
    """All exportable data sections."""

    settings: UserSettings | None = None
    sources: list[ExportedSource] | None = None
    websdrs: list[ExportedWebSDR] | None = None
    sessions: list[ExportedSession] | None = None
    sample_sets: list[ExportedSampleSet] | None = None
    models: list[ExportedModel] | None = None
    audio_library: list[ExportedAudioLibrary] | None = None


class HeimdallFile(BaseModel):
    """Complete .heimdall file structure."""

    metadata: ExportMetadata
    sections: ExportSections


class SampleSetExportConfig(BaseModel):
    """Configuration for exporting a sample set with range selection."""

    dataset_id: str
    sample_offset: int = Field(default=0, ge=0, description="Starting offset for samples")
    sample_limit: int | None = Field(
        default=None, ge=1, description="Number of samples to export (None = all)"
    )
    include_iq_data: bool = Field(
        default=True, description="Include binary IQ data (can be very large, ~16MB per sample)"
    )


class ExportRequest(BaseModel):
    """Request to export data."""

    creator: CreatorInfo
    description: str | None = None
    include_settings: bool = False
    include_sources: bool = True
    include_websdrs: bool = True
    include_sessions: bool = False
    # Sample sets (synthetic datasets) - now with range support
    sample_set_configs: list[SampleSetExportConfig] | None = None
    # Models - list of IDs or None for none
    model_ids: list[str] | None = None
    # Audio library - list of IDs or None for none
    audio_library_ids: list[str] | None = None


class ImportRequest(BaseModel):
    """Request to import data."""

    heimdall_file: HeimdallFile
    import_settings: bool = False
    import_sources: bool = True
    import_websdrs: bool = True
    import_sessions: bool = False
    import_sample_sets: bool = False
    import_models: bool = False
    import_audio_library: bool = False
    overwrite_existing: bool = False


class ExportResponse(BaseModel):
    """Response containing the exported .heimdall file."""

    file: HeimdallFile
    size_bytes: int


class ImportResponse(BaseModel):
    """Response after importing data."""

    success: bool
    message: str
    imported_counts: dict[str, int] = Field(default_factory=dict)
    errors: list[str] = Field(default_factory=list)


class AvailableSampleSet(BaseModel):
    """Available sample set for export."""

    id: str
    name: str
    num_samples: int  # Number of feature samples
    num_iq_samples: int = 0  # Number of IQ samples (raw RF data)
    created_at: str
    estimated_size_bytes: int = Field(
        default=0, description="Estimated size of full dataset in bytes (features + IQ data)"
    )
    estimated_size_per_feature: int = Field(
        default=5600, description="Estimated size per feature sample in bytes"
    )
    estimated_size_per_iq: int = Field(
        default=13_000_000, description="Estimated size per IQ sample in bytes (~13MB)"
    )


class AvailableModel(BaseModel):
    """Available model for export."""

    id: str
    model_name: str
    version: int
    created_at: str
    has_onnx: bool


class AvailableAudioLibrary(BaseModel):
    """Available audio library entry for export."""

    id: str
    filename: str
    category: str
    duration_seconds: float
    total_chunks: int
    file_size_bytes: int  # Original audio file size
    chunks_total_bytes: int  # Total size of preprocessed chunks in MinIO
    created_at: str


class MetadataResponse(BaseModel):
    """Response with available data for export."""

    sources_count: int = 0
    websdrs_count: int = 0
    sessions_count: int = 0
    sample_sets: list[AvailableSampleSet] = Field(default_factory=list)
    models: list[AvailableModel] = Field(default_factory=list)
    audio_library: list[AvailableAudioLibrary] = Field(default_factory=list)
    estimated_sizes: SectionSizes = Field(default_factory=SectionSizes)
